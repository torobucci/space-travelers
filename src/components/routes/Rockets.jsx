import { React } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Button from 'react-bootstrap/esm/Button';
import Badge from 'react-bootstrap/Badge';
import { reserveRocket, cancelReservation } from '../redux/rockects/rocketsSlice';

const Rockets = () => {
  const { rockets } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();
  return (
    <div className="container">
      {rockets.map((rocket) => (
        <div className="d-flex m-4" key={rocket.id}>
          <img src={rocket.flickr_images[0]} alt="rocket-img" width="300px" />
          <div className="ms-4">
            <h1>{rocket.name}</h1>
            <p>
              {rocket.reserved && (
              <Badge bg="primary" className="me-2">
                Reserved
              </Badge>
              )}
              {rocket.description}
            </p>
            {rocket.reserved && (
              <Button variant="outline-secondary" onClick={() => dispatch(cancelReservation(rocket.id))}>Cancel Reservation</Button>
            )}
            {!rocket.reserved && (
              <Button variant="primary" onClick={() => dispatch(reserveRocket(rocket.id))}>Reserve Rocket</Button>
            )}
          </div>
        </div>
      ))}
    </div>
  );
};

export default Rockets;
