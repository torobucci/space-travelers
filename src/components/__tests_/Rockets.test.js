import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import renderer from 'react-test-renderer';
import Rockets from '../routes/Rockets';
import '@testing-library/jest-dom';

jest.mock('react-redux');
jest.mock('react-bootstrap/esm/Button', () => 'MockedBootstrapButton');
jest.mock('axios', () => 'Mockedaxios');

describe('Test Rockets Component', () => {
  const mockRockets = [
    {
      id: 'falcon1',
      name: 'Falcon 1',
      description: 'The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009.',
      flickr_images: ['https://farm1.staticflickr.com/27/60914660_3a1a118583_b.jpg'],
      reserved: true,
    },
    {
      id: 'falcon9',
      name: 'Falcon 9',
      description: 'Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.',
      flickr_images: ['https://farm1.staticflickr.com/745/32394687645_a9c54a34ef_b.jpg'],
      reserved: false,
    },
  ];

  beforeEach(() => {
    useSelector.mockImplementation((selector) => selector({
      rockets: {
        rockets: mockRockets,
      },
    }));
    useDispatch.mockReturnValue(jest.fn());
  });

  it('renders the rocket list', () => {
    const { getByText } = render(<Rockets />);
    expect(getByText('Falcon 1')).toBeInTheDocument();
    expect(getByText('Falcon 9')).toBeInTheDocument();
  });

  it('displays reserved badge when rocket is reserved', () => {
    const { getByText } = render(<Rockets />);
    expect(getByText('Reserved')).toBeInTheDocument();
  });

  it('calls reserveRocket when reserve button is clicked', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    const { getByText } = render(<Rockets />);
    const reserveButton = getByText('Reserve Rocket');
    fireEvent.click(reserveButton);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith({ type: 'rockets/reserveRocket', payload: 'falcon9' });
  });

  it('calls cancelReservation when cancel button is clicked', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    const { getByText } = render(<Rockets />);
    const cancelButton = getByText('Cancel Reservation');
    fireEvent.click(cancelButton);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith({ type: 'rockets/cancelReservation', payload: 'falcon1' });
  });
  it('Rockets snapshots match', () => {
    const tree = renderer.create(<Rockets />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
