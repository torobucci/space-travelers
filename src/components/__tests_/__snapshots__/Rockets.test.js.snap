// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rockets Component Rockets snapshots match 1`] = `
<div
  className="container"
>
  <div
    className="d-flex m-4"
  >
    <img
      alt="rocket-img"
      src="https://farm1.staticflickr.com/27/60914660_3a1a118583_b.jpg"
      width="300px"
    />
    <div
      className="ms-4"
    >
      <h1>
        Falcon 1
      </h1>
      <p>
        <span
          className="me-2 badge bg-primary"
        >
          Reserved
        </span>
        The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009.
      </p>
      <MockedBootstrapButton
        onClick={[Function]}
        variant="outline-secondary"
      >
        Cancel Reservation
      </MockedBootstrapButton>
    </div>
  </div>
  <div
    className="d-flex m-4"
  >
    <img
      alt="rocket-img"
      src="https://farm1.staticflickr.com/745/32394687645_a9c54a34ef_b.jpg"
      width="300px"
    />
    <div
      className="ms-4"
    >
      <h1>
        Falcon 9
      </h1>
      <p>
        Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
      </p>
      <MockedBootstrapButton
        onClick={[Function]}
        variant="primary"
      >
        Reserve Rocket
      </MockedBootstrapButton>
    </div>
  </div>
</div>
`;
